course: task 4.1 decimal calculator

BorislavAnchev opened the pull request:
The decimal calculator task.

BorislavAnchev commented:
Code ready for review.

mgenov commented:
Is this PR going to be re-created or It is supposed to be reviewed as it is?

BorislavAnchev commented:
Task 2.3 was to make a calculator with whole numbers. Task 4.1 is to make it work with decimal numbers. I posted this in the build tools repository, but you made this one and I created the PR here. The code should be according to the comments made about the previous calculator and the tests should be according to the comments about the Counter task.

I'm not sure what 're-created' means, so I guess review as it is? If I had to do something else, tell me.

mgenov commented on Task_4.1/karma.conf.js:
```diff
@@ -0,0 +1,191 @@
+
+// OTHER CONfig
+
+// module.exports = function (config) {
```
This code could be removed. You already have working and reviewed config and you can use it.

BorislavAnchev replied:
Removed.

mgenov commented on Task_4.1/src/calc.js:
```diff
@@ -0,0 +1,70 @@
+  }
+}
+
+export const calc = new Calculator();
```
it would be better the client code to instantiate Calculator on it's own.

BorislavAnchev replied:
But since I use the calculator object here: 

```
buttons.forEach(function(item) {
  document.getElementById(item).addEventListener('click', function() {
    if(/\.|\d|\+|-|X|\//.test(this.value)) {
      if(this.value === 'X') { calc.onInput('*'); }
      else { calc.onInput(this.value); }
    }
    else if(/^C$/.test(this.value)) { calc.onClean(); }
    else if(/</.test(this.value)) { calc.onDelete(); }
    else if(/=/.test(this.value)) { calc.onEquals(); }
    document.getElementById('currentInput').value = calc.currentInput;
    document.getElementById('memoryField').value = calc.memoryField;
  })
```
Should I instantiate it in the eventListeners.js file? And not in the HTML file?

Otherwise I will have 1 instance of the calculator object in eventListeners.js  in order to create the event listeners and another instance in the HTML.

Or should I instantiate the calculator in the HTML and create the event listeners there locally?

mgenov replied:
HTML acts as client of this code, so you can instantiate it in the HTML page.

BorislavAnchev replied:
I can't instantiate the calc object in the HTML, because WebPack injects the JavaScript at the end of the body tag. So, even if I write
`<script> const calc = new Calculator()</script>`
at the end of the body tag, the class declaration will be below it and thus will still be undefined at the moment of the creation of the Calculator instance.

If I write
<script src="./calc"></script>
I'm importing it manually and not using one of WebPack's main functions and also when I create the production build, it won't work, because it will only have 'app.bundle.js' in it's local directory and the correct path will be '../src/calc'.
I also think that I should not be looking for files in the 'src' folder when in the production build, i.e. the production build should have everything it needs in its own folder, right?
Therefore, I must use only 'app.bundle.js' file that WebPack created. WebPack should provide all the needed JavaScript to me.
But that way I cannot instantiate the Calculator in the HTML because it is injected at the end of the client.

Am I looking at it from a completely wrong angle? Maybe I didn't understand you and you wanted me to do something different. If I take the previous calculator, for example, then there will be no problem - import the JS file, then instantiate the Calculator. But there was no WebPack. I was importing it manually and that way there is no problem to put the instantiation after the JavaScript import. But in this task it is said to use WebPack.

So, I may be missing something or I just don't understand you.

mgenov commented on Task_4.1/src/calc.spec.js:
```diff
@@ -0,0 +1,167 @@
+  });
+  
+  describe('.onInput method', () => {
+    it('should change currentInput to 1', () => {
```
doesn't look reasonable method/function to be tested. The goal of the tests is to test behaviour such as 1+3=4

mgenov commented on Task_4.1/src/eventListeners.js:
```diff
@@ -0,0 +1,23 @@
+import { calc } from './calc';
```
import Calculator from './calc'

const calculator = new Calculator();

mgenov replied:
Note that the name of the file doesn't sound good.

BorislavAnchev replied:
So, I should instantiate it here and not in the index.html?

mgenov replied:
Yes, you can do it here.

BorislavAnchev replied:
In that case, the code is ready for another review.

BorislavAnchev commented:
Some changes ready for review.

- Corrected the name of 'calc.js' to 'Calculator.js' to better correspond to its contents.
- Corrected the name of 'calc.spec.js' to 'Calculator.spec.js' to correspond to the file it is written for.
- Corrected the tests, hopefully they will correspond more to testing a behaviour more than testing methods.
- Removed missed out old config.
- Removed the export of the Calculator instance and the instance itself. The instance is now being created in the 'eventListeners.js' file and the Calculator class is being exported instead of it.

Potential pending issues:
- The Calculator class is not yet instantiated in the client side. But since a new correction is made that instantiates it in the 'eventListeners.js' file, it is not clear if this was not the correction you had in mind.

mgenov commented on Task_4.1/src/Calculator.spec.js:
```diff
@@ -0,0 +1,90 @@
+  
+  it('currentInput display value should change from 0 to 1 on initial input of 1', () => {
+    calc.onInput('5');
+    expect(calc.currentInput).to.equal('5');
```
You are using internal of the class to verify state which couples test with the implementation and it increases code maintenance (if you can production code you have to change and the test). 

A better approach is to test the behaviour: 

happy path
```jsx
var calculator = new Calculator()
calc.onInput('5')
calc.onInput('+')
calc.onInput('3')
expect(calc.onEqual()).to.equal('8')
```

concatenates input values
```jsx
var calculator = new Calculator()
var value = calc.onInput('5')
value += calc.onInput('+')
expect(value).to.equal('5+')
```

mgenov replied:
Please check this comment as it provides you with an example of how your tests should look like.

mgenov commented on Task_4.1/src/Calculator.js:
```diff
@@ -0,0 +1,68 @@
+    this.resetInputFlag = resetInputFlag;
+  }
+
+  onInput(input) {
```
Code could become more readable if you split operators from digits. In HTML it's easy as these are separate buttons.

BorislavAnchev commented:
New changes ready for review:

- onInput() method is split into onOperator() and onDigit() methods.
- The calculator no longer uses its internal fields to store the currentInput and memoryField. Instead, the methods now return the new display values for the both fields.
- The tests are reworked according to the given example for behaviour testing.

mgenov commented on Task_4.1/src/Calculator.js:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  onOperator(input, currentInput, memoryField) {
+    if(input === '.') {
```
Not sure that "." is part of the operator.

mgenov commented on Task_4.1/src/Calculator.spec.js:
```diff
@@ -0,0 +1,62 @@
+const calc = new Calculator();
+
+describe('The calculator object', () => {
+  it('should have his resetInputFlag set to true on initialization', () => {
```
What is the goal of this test?

mgenov commented on Task_4.1/src/Calculator.spec.js:
```diff
@@ -0,0 +1,62 @@
+  
+  it('should produce 8 as the result of 3+5 and set the resetInputFlag to true', () => {
+    calc.resetInputFlag = false;
+    expect(calc.onEquals('3+', '5')).to.equal(8);
```
onEquals should be called when expression is entered and looks strange this function to accept arguments.

BorislavAnchev commented:
New code is ready for review:
- onEquals() now accepts no arguments. This meant that the calculator had to keep its internal state for the calculation of the expression.
- Initialization test at the beginning is removed.
- The decimal point is moved into the onDigit() method.
- Test descriptions are made more descriptive and assertions for the internal fields are added.

mgenov commented on Task_4.1/src/Calculator.spec.js:
```diff
@@ -0,0 +1,77 @@
+const calc = new Calculator();
+
+describe('The calculator object', () => {
+  it('should produce 8 as the result of 3+5, set the resetInputFlag to true, reset the memoryField and set the currentInput to the result', () => {
```
The test is still using the internals of the Calculator class.

mgenov commented on Task_4.1/src/Calculator.spec.js:
```diff
@@ -0,0 +1,77 @@
+    expect(calc.currentInput).to.equal('3.5');
+  });
+
+  it('should append the decimal point to a zero, if the resetInputFlag is true, set the currentInput to "0.", and set the resetInputFlag to false', () => {
```
The name of the scenario doesn't sound good as it's coupled to concrete internal of the Calculator class. Does this mean that If resetInputFlag is renamed to something else in the production code all tests have to be renamed too?

BorislavAnchev commented:
New code is ready for review:
- Removed usage of class internals from the tests.
- Rewrote the test scenarios in a more general way.
- Added an afterEach() function in order to reset the state of the object after each test.
